kliff/calculators/calculator_torch.py

```diff
@@ -253,6 +276,89 @@ def _compute_stress(denergy_dzeta, dzetadr, volume):
         return forces
 
+    def get_size_opt_params(self):
+        """
+        Return the size of the parameters.
+
+        Returns
+        -------
+        sizes: list
+            Each element in the list gives the shape of each type of parameter tensors,
+            containing, e.g., weights and biases, for each layer.
+        nelements: list
+            Number of elements of each parameter tensor.
+        nparams: int
+            Total number of parameters
+        """
+        sizes = []  # Size of each parameter tensor
+        nelements = []  # The number of elements for each tensor
+        for param in self.model.parameters():
+            sizes.append(param.size())
+            nelements.append(np.prod(param.size()))
+        nparams = sum(nelements)
+        return sizes, nelements, nparams
+
+    def get_num_opt_params(self):
+        """
+        Return the total number of parameters.
+        """
+        return self.get_size_opt_params()[-1]
+
+    def get_opt_params(self, flat=True):
+        """
+        Retrieve the parameters, i.e., weights and biases.
+
+        Parameters
+        ----------
+        flat: bool (optional)
+            A flag to return a flat, 1D array.
+
+        Returns
+        -------
+        list or np.ndarray
+            Parameters, i.e., weights and biases. If ``flat=True``, a 1D np.ndarray will
+            be returned. Otherwise, nested lists will be returned, where each list contain
+            the weights and biases for each layer.
+        """
+        parameters = []
+        for param in self.model.parameters():
+            parameters = np.append(parameters, param.data.numpy().flatten())
+        if flat:
+            # Return a 1D array
+            return parameters
+        else:
+            # Return a list of tensors
+            return self._convert_from_flat_parameters(parameters)
+
+    def update_model_params(self, parameters):
+        """
+        Update the model parameters from a 1D array.
+
+        Parameters
+        ----------
+        parameters: np.ndarray
+            New parameter values to set. It needs to be a 1D array.
+        """
+        # Convert to the right format
+        parameters = self._convert_from_flat_parameters(parameters)
+        # Update the weights and biases
+        for ii, param in enumerate(self.model.parameters()):
+            param.data = parameters[ii]
+
+    def _convert_from_flat_parameters(self, flat_params):
+        """
+        Convert the parameters from a 1D array format to nested lists format.
+        """
+        sizes, nelems, _ = self.get_size_opt_params()
+        # Indices to index the flat array to get the appropriate portion of each parameter
+        # tensor
+        idx = np.append(0.0, np.cumsum(nelems)).astype(int)
+        parameters = []
+        for ii, size in enumerate(sizes):
+            params = flat_params[idx[ii] : idx[ii + 1]]
+            parameters.append(torch.Tensor(params.reshape(size)))
```

Comment:
The parameters are going to be used only at reference time, i.e. for predictions? right? Then this is probably OK. 

But I still recommend replacing `torch.Tensor` by `torch.Parameter` https://pytorch.org/docs/stable/generated/torch.nn.parameter.Parameter.html. The latter is a special tensor that requires grad, which is necessary of we want to take automatic differentiation w.r.t. the parameters.

Reply:
Currently, I only need this to retrieve the parameters to export the samples, and when we want to use the samples to update the model parameters and use the model for other calculations. But, I can see that your suggestion is more future-proof.

Reply:
I changed `torch. Tensor` to `torch. Parameter`, but I got some unwanted behavior in `update_model_parameters`. When I changed to use `torch. Parameter`, I changed the line `param.data = parameters[ii]` to `param = parameters[ii]`, where `parameters[ii]` now is `torch. Parameter`. However, from my testing, this doesn't change the parameters. I think the way to update the parameters should be by using `param.data = <torch. Tensor>`. Because of this, I will revert back the change. I will also add a test for this functionality.

Reply:
My bad. What you are doing is 100% correct. I missed that you are assigning this back to `tensor.data` in my new comment above. Sorry about that!